v2/datastream-to-sql/src/main/java/com/google/cloud/teleport/v2/utils/DatastreamToPostgresDML.java

```diff
@@ -102,6 +107,56 @@ public String cleanDataTypeValueSql(
         }
         break;
     }
+    // Arrays in Postgres are prefixed with underscore e.g. _INT4 for integer array.
+    if (dataType.startsWith("_")) {
+      return convertJsonToPostgresArray(columnValue);
+    }
     return columnValue;
   }
+
+  private String convertJsonToPostgresArray(String jsonValue) {
+    if (jsonValue == null || jsonValue.equals("''") || jsonValue.equals("")) {
+      return getNullValueSql();
+    }
+
+    try {
+      ObjectMapper mapper = new ObjectMapper();
+      JsonNode rootNode = mapper.readTree(jsonValue);
+
+      if (!(rootNode.isObject() && rootNode.has("nestedArray"))) {
+        LOG.warn("Empty array: {}", jsonValue);
+        return getNullValueSql();
+      }
+
+      JsonNode arrayNode = rootNode.get("nestedArray");
+
+      // Handle nested structure with elementValue
+      List<String> elements = new ArrayList<>();
+      if (arrayNode.isArray()) {
+        for (JsonNode element : arrayNode) {
+          if (element.has("elementValue")) {
+            JsonNode elementValue = element.get("elementValue");
+            if (!elementValue.isNull()) {
+              elements.add(formatArrayElement(elementValue));
+            } else {
+              elements.add(getNullValueSql());
+            }
+          } else if (!element.isNull()) {
+            elements.add(formatArrayElement(element));
+          }
+        }
+      }
+      return "ARRAY[" + String.join(",", elements) + "]";
```

Comment:
why are we adding ARRAY here?  shouldn't it just be the value?

Reply:
The CL is formatting this data directly into Postgres syntax but it should be either string json or an object at this point.  In the Postgres writer you'll want to wrap thi value in the array syntax (the file i sent earlier)
https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/main/v2/datastream-to-sql/src/main/java/com/google/cloud/teleport/v2/utils/DatastreamToPostgresDML.java#L103

Reply:
apologies, nvm, i had thought we were doing this in the avro reader ... ignore